core/vote/vote_manager.go

```diff
@@ -34,7 +32,38 @@ type VoteManager struct {
 
 	engine consensus.PoSA
 
-	getJustifiedHeader getJustifiedHeaderFunc
+	voteDataBuffer *voteDataBuffer
+}
+
+type voteDataBuffer struct {
+	voteDataMap map[uint64]*types.VoteData
+	cache       []uint64
+}
+
+func newVoteDataBuffer() *voteDataBuffer {
+	return &voteDataBuffer{
+		voteDataMap: make(map[uint64]*types.VoteData, maxSizeOfRecentEntry),
+		cache:       make([]uint64, 0, maxSizeOfRecentEntry),
+	}
+}
+
+func (voteDataBuffer *voteDataBuffer) add(voteData *types.VoteData) {
```

Comment:
I only see one place call this add function, it is not enough, we should make sure the map is definitely the same as the vote journal, we should handle restart case also. 

As the vote map needs to sync with the vote journal, I think integrating voteDataBuffer as a part of the Vote Journal will be better. The Vote Manager no need to call this add function, we can hide this add operation within the WriteVote function of the vote journal.

Reply:
Resolved